### DISCUSSION TITLE
Example changes to be made to use nvs example or read write to internal flash on STM32H750B-DK

### DISCUSSION THREAD

**Discussion by MaazSk** [zephyrproject-rtos/zephyr#75997] [Q&A] [answered]
I am using STM32H750B-DK board I want to run the nvs example on it and need help regarding setting up the overlay file for the board. I am new to zephyr and unable to figure out what changes are required to make it work

**Comment by nordicjm** [comment]
NVS will not work on that device due to the sector size

**Reply by MaazSk** [reply]
okay, is there a example for the same board to write to flash?

**Reply by nordicjm** [reply]
https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/flash/common

**Reply by MaazSk** [reply]
I tried that but it does not work for my board it gives error even I tried stm32 tests

**Reply by Laczen** [reply]
@MaazSk In theory your board should be capable of supporting nvs. However your board has some peculiarities that might make it challenging to support the flash API. Your board is using a "dual" external flash setup (at least the latest in tree dts) that has received little testing. While looking at the driver I think the page-layout is not correctly reflecting the "dual" flash (@FRASTM doesn't the erase-block needs to be doubled for "dual" flash). Next to that it is also not clear where to put the flash partition layout in the dts for a "dual" flash layout. I would advice to enable the flash shell (CONFIG_FLASH_SHELL=y) and to use the shell to test the flash (find out the flash device, test erase, test write, ...). You will need to be forgiving about a possible driver error (e.g. test with erase size = 8kB instead of the expected 4kB). Test writing/erasing somewhere at a 8kB border past your image size. Once you have some basic idea that the flash is working (and you know the device name) it is possible to look at partitioning the flash.

**Reply by MaazSk** [reply]
Okay, I will give flash_shell a try and if there's anything I will let you know. Thank you!

**Reply by MaazSk** [reply]
@Laczen it is working flash write erase is working I'll move forward now. What I am trying now is that I won't have the partition but will write a code to write and read from internal flash. Is there a way to get the device name directly. Sorry for asking so many questions but I am new to zephyr.

**Reply by nordicjm** [reply]
> @MaazSk In theory your board should be capable of supporting nvs. However your board has some peculiarities that might make it challenging to support the flash API. Your board is using a "dual" external flash setup (at least the latest in tree dts) that has received little testing. While looking at the driver I think the page-layout is not correctly reflecting the "dual" flash (@FRASTM doesn't the erase-block needs to be doubled for "dual" flash). Next to that it is also not clear where to put the flash partition layout in the dts for a "dual" flash layout. I would advice to enable the flash shell (CONFIG_FLASH_SHELL=y) and to use the shell to test the flash (find out the flash device, test erase, test write, ...). You will need to be forgiving about a possible driver error (e.g. test with erase size = 8kB instead of the expected 4kB). Test writing/erasing somewhere at a 8kB border past your image size. Once you have some basic idea that the flash is working (and you know the device name) it is possible to look at partitioning the flash.

How will this work, the sector size is 128KiB? You set a maximum size of 64KiB

**Reply by Laczen** [reply]
What you have probably been testing is the internal flash (flash0). On your device the internal flash is very limited and meant to be used for a bootloader that starts a program from external flash. Although you could use a part of the internal flash for nvs it is better to focus on trying to add storage (either nvs or later littlefs) on the external device.
Using the shell you can find out which devices are available ("device list") that should allow retrieving the external flash device name. You can then test the read/write/erase on the external flash.

**Reply by Laczen** [reply]
> 
> How will this work, the sector size is 128KiB? You set a maximum size of 64KiB

The device/development kit has (dual) external flash. The external flash will allow a erase-block of 4kB (but as both flashes are used at the same time, even data stored in external flash1 odd data in external flash2) the "real" erase block will be 8kB. The 128kB internal flash seems to be indeed only 1 "sector" so it can't really be used for anything but a bootloader.
How to setup the dts with dual flash (partitioning) is not clear to me yet.

**Reply by MaazSk** [reply]
I am still not sure how to simply write and read from internal flash is there a sample code available that can be used?
zephyr/tests/drivers/flash/stm32 passes the test but how can I use it in my code the .dts file generated in the build folder is different from other .dts file specially the flash part is there a sample code to read and write in internal flash?

**Comment by FRASTM** [comment]
For the stm32h750 , the internal flash is 128KB containing _one_ user sector fo 128KB
this sector size = 0x20000 exceeds the `uint16_t sector_size; 	/** Number of sectors in the file system */`
of the `struct nvs_fs {`
	
The samples/subsys/nvs/ cannot run on the internal flash memory since the (single) sector used for storage also contains the code. It cannot be erased !

**Reply by MaazSk** [reply]
Okay, is there a simple read and write to internal flash sample code available? I tried zephyr/tests/drivers/flash/stm32 passes the test but how can I use it in my code Also the .dts file generated for the test has a storage_partition whereas when I run a normal code it does not have a storage_partition in it. 

Is there a sample code to read and write in internal flash?

**Reply by Laczen** [reply]
@MaazSk you should not look at the internal flash, it is unusable for storage as it only has 1 (large) sector. It is the external flash that can be used for storage. The storage_parttion should be defined on the external flash. The problem is that the external flash is made up of 2 (q)spi flashes and the partition needs to be added to one/both of these to allow flash_map to retrieve the device it is working on. Your dk is the first (AFAIK) that has this dual external flash setup so it is the first to have this problem.

**Reply by MaazSk** [reply]
@Laczen Currently what I am trying and want to do the POC is just to read and write to the internal flash. I have a small amount of data say an array of 50 bytes which I want to store in the internal flash and read it whenever necessary. So, that's the reason I am focusing on reading and writing to internal flash. 
Also, my next step is to have flash protection feature

**Reply by Laczen** [reply]
> @Laczen Currently what I am trying and want to do the POC is just to read and write to the internal flash. I have a small amount of data say an array of 50 bytes which I want to store in the internal flash and read it whenever necessary. So, that's the reason I am focusing on reading and writing to internal flash. Also, my next step is to have flash protection feature

You can do that, but... The internal flash is only 1 sector, if your firmware is running from this location the data will be removed everytime you update your firmware. If your firmware is running from external flash you can only use this with the flash interface, but not to create "persistent" storage because whenever you need to erase the flash block it will erase everything that is there. It should be straight forward to program this with the flash api, but it will have limitations as a result of the 1 sector.

**Comment by FRASTM** [comment]
write/reading the internal flash of the `stm32h750b_dk` out of the code area is still possible with the   samples/drivers/flash_shell/
once the sector has been erased previously (it was during the west flash), but it is like an OTP
```
*** Booting Zephyr OS build v3.7.0-rc1-589-g85a0440af172 ***                                                          
uart:~$ flash write flash-controller@52002000 0x1f000 0x55555555                                                      
Write OK.                                                                                                             
Verified.    
```

**Reply by MaazSk** [reply]
I did this and was able to write, read and erase but how to do the same using code?

**Reply by FRASTM** [reply]
> I did this and was able to write, read and erase but how to do the same using code?

I suggest to build a basic application (model could be tests/drivers/flash/common) with flash_write() and flash_read() on the flash_dev device

**Reply by MaazSk** [reply]
okay will try this just give me some time

**Reply by FRASTM** [reply, marked as answer]
take a look a this 
[flash_h7.zip](https://github.com/user-attachments/files/16283755/flash_h7.zip)
west build -p always -b stm32h750b_dk samples/boards/stm32/flash_h7/
```
*** Booting Zephyr OS build v3.7.0-rc1-588-g40810983ead2 ***

flash-controller@52002000 STM32H7 flash testing
==========================

Test 1: check Flash erased
Flash erase succeeded!

Test 2: Flash write
Attempting to write 16 bytes
Data read matches data written. Good!!
```

**Reply by MaazSk** [reply]
Hi @FRASTM,
Thank you! for the code it works fine and gives me a basic idea of how to read and write.
I have a small question when I reset the device it gives me an error that it is unable to write and the error code is -5. Can you please help me why is that so? Also, I made a minor change in the code attaching the same for your reference

> Test 2: Flash write
> Attempting to write 16 bytes
> Flash write failed! -5

[flash_h7.zip](https://github.com/user-attachments/files/16321169/flash_h7.zip)

**Reply by FRASTM** [reply]
> Hi @FRASTM, Thank you! for the code it works fine and gives me a basic idea of how to read and write. I have a small question when I reset the device it gives me an error that it is unable to write and the error code is -5. Can you please help me why is that so? Also, I made a minor change in the code attaching the same for your reference
> 
> > Test 2: Flash write
> > Attempting to write 16 bytes
> > Flash write failed! -5
> 
> [flash_h7.zip](https://github.com/user-attachments/files/16321169/flash_h7.zip)

Resetting the board does not erase the flash. It can be written only one time after erasing. 
So after flashing (which erases the sector) and first run, any next run or next tentative to write (like executing this sample) will fail. 
As said this flash area is more an OTP.

**Reply by MaazSk** [reply]
Hi @FRASTM,

One last thing I want to confirm so I wanted a write protection so if it is something like an OTP do I need to implement a write protect?
I was trying to add 
```
uintptr_t code = (uintptr_t)OB_WRPSTATE_ENABLE;
	rc = flash_ex_op(flash_dev, OPTIONBYTE_WRP, code, NULL);
	if (rc != 0)
	{
		printf(" Write Protection failed with error: %d", rc);
	}
```

But it failed everytime and gave me an error `-88`
Also, I tried to change it to 
```
uintptr_t code = (uintptr_t)OB_WRPSTATE_ENABLE;
	rc = flash_ex_op(flash_dev, OPTIONBYTE_WRP, NULL, NULL);
	if (rc != 0)
	{
		printf(" Write Protection failed with error: %d", rc);
	}
```

But the output was same so I wanted to confirm do I need to implement a write protect or not?
Also, what is the correct way to implement the api `flash_ex_op`

Thank you in advance

**Reply by Laczen** [reply]
It is only a OTP because of your usage pattern (the image and data reside in the same flash erase block). When the image is updated the data that you have written is erased. There is no need to implement some write protection, it is not allowed to rewrite an already written location. You can check if an area is erased before writing any data to it.

Although you can use this for some storage over reboot it has very limited usability. It would be good to share the intended usage so we can help in giving the right solution direction.

**Reply by FRASTM** [reply]
> It is only a OTP because of your usage pattern (the image and data reside in the same flash erase block). When the image is updated the data that you have written is erased. There is no need to implement some write protection, it is not allowed to rewrite an already written location. You can check if an area is erased before writing any data to it.
> 
> Although you can use this for some storage over reboot it has very limited usability. It would be good to share the intended usage so we can help in giving the right solution direction.

I agree. 
@MaazSk you can still add any comment or LOG  to explain the particular situation of the stm32h750

**Reply by MaazSk** [reply]
@Laczen and @FRASTM I am trying to do a POC for write protect as my application needs to store some unique identification number for each product. I am going to write this number once only into the flash and was thinking to do a write protect which will prevent overwriting of that area. Earlier I used OTP area on stm32u575 and it worked fine as it was protected. Is there any way to this?

**Reply by Laczen** [reply]
@MaazSk the device you are working with is very flexible and highly customizable. The "intended" use of this device is IMHO to add a "small" bootloader (this can be a small zephyr application e.g. mcuboot) to the 128kB internal flash that starts an application from flash by either copying it to ram (while optionally decrypting) or by starting it directly from external flash in memory mapped mode. When the "small" bootloader is small enough the remaining area can be used for OTP (as long as the "small" bootloader is not replaced with something else). As it is flash it will not allow you to overwrite data what you have experienced in your setup. Your application can read the data, but when you enable the flash driver to read it might also be possible to erase it. To avoid this and to make sure that nothing happens to the "small" bootloader I would take a slightly different approach:
1. an area in flash0 is reserved to store never changing values (like in the flash_h7.zip),
2. the small bootloader reads this while starting and copies it to a ram memory area (a section of ram dedicated for this),
3. for the application the flash0 device is disabled, so that it is not possible to read/write or erase. The application gets the data from the ram area.

BTW I saw in the datasheet that the device also has a 2kB area for user option bytes, so this might also be an option.

**Reply by MaazSk** [reply]
@Laczen can you please help me with what option bytes are or probably the section from the reference manual would be great I'll check it and then if anything is required I will let you know.

Thanks

**Reply by Laczen** [reply]
It seems that the user option bytes are really flash option bytes, that can't be used to store any data. More information can be found in section 4.4 of the reference manual.

**Comment by FRASTM** [comment]
![Screenshot from 2024-07-23 16-10-50](https://github.com/user-attachments/assets/f37745d7-f6aa-4d3c-9e0a-0914e2d8b808)
